docs/reference/style-guides/node-style-guide.md

```diff
@@ -2,8 +2,311 @@
+
+1. Check if the variable is null/undefined, and if so throw an error
+
+Do:
+
+```typescript
+function example(arg?: string) {
+  if (!arg) {
+    throw new Error('arg is required for example');
+  }
+
+  // arg is now of type 'string'
+}
+```
+
+Don't:
+
+```typescript
+function example(arg?: string) {
+  return arg!.substring(0, 20);
+}
+```
+
+2. Return early if it matches the desired behavior
+
+Do:
+
+```typescript
+function example(arg?: string) {
+  if (!arg) return;
+
+  // arg is now of type 'string'
+}
+```
+
+3. Fallback to a default
+
+Do:
+
+```typescript
+const DEFAULT_VALUE = 'example';
+
+function example(arg?: string) {
+  const argWithDefault = arg || DEFAULT_VALUE;
+
+  // argWithDefault is of type 'string'
+}
+```
+
+4. Narrow the type with conditions
+
+Do:
+
+```typescript
+interface MyType {
+  prop: string;
+}
+interface AnotherType {
+  prop2: string;
+}
+function otherFunc(): MyType | AnotherType;
+
+function example(): MyType {
+  const result = otherFunc();
+
+  if (!('prop' in result)) throw new Error('result was of an unexpected type');
+
+  // result is now of type MyType
+  return result;
+}
+```
+
+Don't:
+
+```typescript
+function example(): MyType {
+  const result = otherFunc();
+  return result as unknown as MyType;
+}
+```
+
+#### Use `satisfies`
+
+Prior to TypeScript 4.9, new literal objects could be identified as a type, but would not be narrowed. Our code uses TypeScript 4.9 or greater, so [`satisfies`](https://www.typescriptlang.org/docs/handbook/release-notes/typescript-4-9.html#the-satisfies-operator) should be preferred.
+
+Don't:
+
+```typescript
+interface Vibe {
+  mood: 'happy' | 'sad';
+}
+
+const vibe: Vibe = {
+  mood: 'happy',
+  // (property) Vibe.mood: "happy" | "sad"
+};
+
+vibe.mood;
+// (property) mood: "happy" | "sad"
+```
+
+Do:
+
+```typescript
+interface Vibe {
+  mood: 'happy' | 'sad';
+}
+
+const vibe = {
+  mood: 'happy',
+  // (property) Vibe.mood: "happy" | "sad"
+} satisfies Vibe;
+
+vibe.mood;
+// (property) mood: "happy"
+```
+
+See [additional examples of using `satisfies`](https://www.learningtypescript.com/articles/the-satisfies-operator).
+
+## Legacy Code Packages
+
+### Style
+
+Legacy code has no defined style-guide beyond that enforced by ESLint.
+
+[ESLint][1] configuration is provided for each legacy package which enforces its requirements, it will be run by default during `git commit`.
+
+[Prettier][2] is used to format the code.
+
+### Structure
+
+A structural convention separates route handling code from library code by prefixing the route handlers with `routes`.
+
+Tests are located in a file hierarchy that mirrors the code being tested, for example a file in a package called:
+
+`fxa-auth-server/lib/bounces.js`
+
+Has a corresponding test at:
+
+`fxa-auth-server/tests/local/bounces.js`
+
+### Async/Await
+
+Some legacy code packages may still have callbacks and/or promises in use. New code should use async/await syntax.
```

Comment:
I've also seen quite a few spots where there will be an `async` operation in a constructor. We should definitely disallow this in all code going forward. I am not sure where this fits into the doc exactly. Perhaps noting it here is fine.

Reply:
Yep, we should definitely avoid that.